rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+Also, SO search functionality is quite limited and doesn’t allow using complex search queries. 
+
+### RemoteRepository
+Cloud plugin registered RemoteRepository as a strategy Kibana Notification Service for creation, storage,
+and retrieval notifications from the remote Unified Notification Service. 
+It’s still not clear what functionality is supported by UNS, but we expect it to be:
+- Create a notification for a particular recipient, a group of recipients. 
+- Read a list notification for a particular recipient.
+- Filter notification list.
+- Change a notification state.
+
+Kibana-specific entities shouldn’t leak to UNS: 
+- Kibana sends message translations for all the supported locales.
+- Kibana-specific groups of recipients unfolded to a list of *recipient_id*.
+- *recipient_id* is set to *elastic_id*.
+
+## Delivery
+The delivery mechanism is responsible for the notifications shipment to Kibana UI. 
+
+#### List of notifications
+For simplicity, we can start with Kibana UI using the HTTP polling mechanism to fetch the last notifications from `/notifications` endpoint.
+Server orders notifications by *pinned_status* and *created_at* timestamp.
+
+Kibana HTTP API:
+- Endpoint: `GET /notifications/` 
+    - `search_after?: number` - Unix timestamp to start searching from.
+    - `type_id?: string[]` - list of source types to filter by.
+    - `size?: number` - max number of notifications to retrieve.
+
+*search_after* and *size* params used to implement pagination logic.
+
+In one of the next phases of the project, we will have to do load testing and decide whether to consider the alternative
+to HTTP polling delivery mechanisms: passing information with a response header (as done for licensing), WebSockets, ServerSideEvents.
+
+#### Notification Status
+The delivery mechanism handles Notification status change performed by Kibana UI.
+Every notification has a different status for every user.
+Supported notification statuses:
+- `read: boolean`
+- `pinned: boolean`
+
+Kibana UI changes notification status to `read: true` when a user interacts with it: clicks on CTA or `Mark as Read` button.
+
+The user can change the notification status to `read: false` manually in the Kibana UI.
+
+The user can change the status for all the existing notifications (`mark all as read`), even for those not-rendered in UI yet.
+
+Kibana HTTP API:
+- mark notification as read
+```
+POST /notification/read
+{ notification_id: string[] }
+```
+- Mark all the notification read
+```
+POST /notifications/_mark_all_read
+```
+- Pin a notification.
+```
+POST /notification/pin
+{ notification_id: string[] }
+```
+- Unpin a notification.
+```
+POST /notification/unpin
+{ notification_id: string[] }
+```
+
+Writing to the same Notification storage might cause conflict on the `write`. If the storage mechanism implementation
+doesn’t make them impossible, we will have to use a single place to change notification status (via the [queue](#localrepository)).
```

Comment:
Are you saying that we should leverage the ES-backed queue to determine the order of state changes? So the flow would be:
- NotificationService.markAsRead() is called
  - NotificationEvent (or similar) is put into ES queue
  - `markAsRead` returns (even before the change has been fully processed)
- Later on, the background task picks up events form the queue:
  - Take the latest X events in queue and process them serially

This should eliminate write conflicts because only one instance of the background task is running (a guarantee provided by Task Manager).

Reply:
> Are you saying that we should leverage the ES-backed queue to determine the order of state changes?

It depends on the data storage implementation. If we end up storing a notification state object per user, the probability of writing conflicts is reduced significantly.